go/libraries/doltcore/sqle/commit_hooks.go

```diff
@@ -292,3 +278,152 @@ func RunAsyncReplicationThreads(bThreads *sql.BackgroundThreads, ctxF func(conte
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a, newThreads := NewAsyncPushOnWriteHook(tmpDir, logger)
+	p := NewPushOnWriteHook(tmpDir, logger)
+
+	if remote != "" {
+		destDb, err := getDestinationDb(ctx, dEnv, remote)
+		if err != nil {
+			return nil, nil, err
+		}
+		p.destDb = destDb
+		a.destDb = destDb
+	}
+
+	return &DynamicPushOnWriteHook{
+		dEnv:      dEnv,
+		tempDir:   tmpDir,
+		logger:    logger,
+		remote:    remote,
+		async:     async,
+		syncHook:  *p,
+		asyncHook: *a,
+	}, newThreads, nil
+}
+
+func getReplicationVals() (string, bool, error) {
+	_, val, ok := sql.SystemVariables.GetGlobal(dsess.ReplicateToRemote)
+	if !ok {
+		return "", false, sql.ErrUnknownSystemVariable.New(dsess.ReplicateToRemote)
+	}
+	remoteName, ok := val.(string)
+	if !ok {
+		return "", false, sql.ErrInvalidSystemVariableValue.New(val)
+	}
+
+	async := false
+	if _, val, ok = sql.SystemVariables.GetGlobal(dsess.AsyncReplication); ok && val == dsess.SysVarTrue {
+		async = true
+	}
+
+	return remoteName, async, nil
+}
+
+func getDestinationDb(ctx context.Context, dEnv *env.DoltEnv, remoteName string) (*doltdb.DoltDB, error) {
+	remotes, err := dEnv.GetRemotes()
+	if err != nil {
+		return nil, err
+	}
+
+	rem, ok := remotes.Get(remoteName)
+	if !ok {
+		return nil, fmt.Errorf("%w: '%s'", env.ErrRemoteNotFound, remoteName)
+	}
+
+	destDb, err := rem.GetRemoteDB(ctx, types.Format_Default, dEnv)
+	if err != nil {
+		return nil, err
+	}
+	return destDb, nil
+}
+
+func (m *DynamicPushOnWriteHook) Execute(ctx context.Context, ds datas.Dataset, db *doltdb.DoltDB) (func(context.Context) error, error) {
+	remoteName, async, err := getReplicationVals()
+
+	// We only need the lock if the remote configuration has changed.
+	hook, err := func() (doltdb.CommitHook, error) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		if m.remote == remoteName && m.async == async {
+			// No change in config since last execution.
+			if m.remote == "" {
+				// replication disabled
+				return nil, nil
+			}
+
+			if async {
+				return &m.asyncHook, nil
+			}
+			return &m.syncHook, nil
+		}
+
+		if remoteName == "" {
+			// replication disabled
+			m.remote = ""
+			m.async = false
+			m.asyncHook.destDb = nil
+			m.syncHook.destDb = nil
+			return nil, nil
+		}
+
+		m.remote = remoteName
+
+		destDb, err := getDestinationDb(ctx, m.dEnv, m.remote)
+		if err != nil {
+			return nil, err
+		}
+
+		m.syncHook.destDb = destDb
+		m.asyncHook.destDb = destDb
+
+		if async {
+			return &m.asyncHook, nil
+		}
+		return &m.syncHook, nil
+	}()
+
+	if err != nil {
+		logrus.Warnf("replication hook failed: %v", err)
+		return nil, err
+	}
+
+	if hook == nil {
+		// replication disabled
+		return nil, nil
+	}
+
+	logrus.Debugf("replication hook invoked. pushing to '%s' (asyn=%t)", remoteName, async)
+	return hook.Execute(ctx, ds, db)
+}
+
+func (m *DynamicPushOnWriteHook) ExecuteForWorkingSets() bool {
+	return false
+}
+
+var _ doltdb.CommitHook = (*DynamicPushOnWriteHook)(nil)
```

Comment:
(minor) most of our code puts these checks right after the struct definition. Nice to see upfront what interfaces a type implements.